### ISSUE TITLE
Tracking issue for `bool::then_some`

### ISSUE THREAD

**Issue by varkor** [rust-lang/rust#64260]
Tracking issue for the `bool::then_some` method, which abstract the following pattern:

```rust
if some_condition {
    Some(t)
} else {
    None
}
```

`bool:then` has previously been stabilised as part of this feature gate: https://github.com/rust-lang/rust/pull/79299

RFC: https://github.com/rust-lang/rfcs/pull/2757
Implementation PR: https://github.com/rust-lang/rust/pull/64255

**Comment by U007D** [comment]
In trying this out, I was puzzled why the eager/lazy argument evaluation versions are `then`/`then_with` and not `and`/`and_then`, which is more consistent with established patterns in `std`?

**Comment by varkor** [comment]
There's a long discussion in [the RFC](https://github.com/rust-lang/rfcs/pull/2757) about the naming. The current names are essentially placeholders.

**Comment by jplatte** [comment]
As a datapoint, this feature turns out pretty useful in proc macros, and I do think the current naming reads nicely: https://github.com/ruma/ruma-api/commit/cb00188e0371c544c4b8622d5a32731f2f6c54b3

**Comment by varkor** [comment]
@jplatte: that would be helpful to share on the RFC!

**Comment by c410-f3r** [comment]
It's been sometime since the PR implementation.
Is there anything preventing stabilization? If not, I can write a stabilization report for this feature.

**Comment by varkor** [comment]
@c410-f3r: the names were only decided upon quite recently. It was decided that we would wait a few versions before deciding to stabilise: https://github.com/rust-lang/rfcs/pull/2757#issuecomment-564683344.

**Comment by U007D** [comment]
Just came back to give this a test drive in a library I am working on.  If we look at common current established patterns in `std`:

`Option`:
`map<U, F> where `F: FnOnce(T) -> U` and
`and_then<U, F> where `F: FnOnce(T) -> Option<U>`

`Result`:
`map<U, F> where `F: FnOnce(T) -> U` and
`and_then<U, F> where `F: FnOnce(T) -> Option<U>`

Looking at `bool`:
`then_some` where `F: FnOnce(T) -> U`

The inconsistency is jarring.  Specifically,
1) for consistency shouldn't the `F: FnOnce() -> U` be called `map`, and `F: FnOnce() -> Option<U>` be a `then` derivative?
2) shouldn't `bool_to_option` also provide *both* an `F: FnOnce() -> U` and an `F: FnOnce() -> Option<U>` mapping?

This is not just theoretical; I am working on a library with a type that has a two-step fallible constructor.  The intent is to be able to write something along the lines of:

```rust
...
impl Foo {
    fn new() -> Option<Self> {
        step_1_test().and_then(|| step_2_test().map(|| Self { /* ... */ }))
    }
}
```

**Comment by KrishnaSannasi** [comment]
You can call `bool::then` followed by `Option::flatten` to get the same effect. But it does make sense to have a fallible version of `bool::then`

**Comment by jplatte** [comment]
> The inconsistency is pretty jarring/unexpected, to my mind.

@U007D: Your comparison is between `bool → Option<T>` conversion functions and `Option<T> → Option<U>` (+ `Result<T, E> → Result<U, E>`) conversion functions. Personally I find that given the input and output types are completely distinct in the new functions when they're not in the ones you're comparing them with is a good reason to consider different names.

**Comment by Boscop** [comment]
@U007D I completely agree. It was already tripping me up many times when using the `boolinator` crate, which also has inconsistent naming!
IMO these functions should be named consistently with `Option`/`Result`.
And not just `map` and `and_then`, but it would also make sense to have `and`, `or` and `or_else`.

Or (if that's too much to include into std): 
Just have a method on `bool` that turns a bool into an option `fn opt(self) -> Option<()>;` (should have a short name), so that we can then use all the above option methods:
`b.opt().map(|_| 42)`, `b.opt().and_then(|_| Some(42))`, `b.opt().and(Some(42))`, `b.opt().or(Some(42))`, `b.opt().or_else(|| Some(42))`.
I'd also prefer to have these as direct methods on `bool`. But please use consistent naming!

**Comment by djc** [comment]
How is the libs team feeling about this? Just ran into another case where I'd have liked to use this (on stable).

**Comment by varkor** [comment]
cc @rust-lang/libs regarding naming and stabilisation: https://github.com/rust-lang/rust/issues/64260#issuecomment-596187828.

**Comment by SimonSapin** [comment]
@varkor Did you mean to link some other comment? This one does not discuss naming.

I honestly don’t remember the current status of this API.

> The current names are essentially placeholders.

Is this still the case or has some consensus been found since?

**Comment by varkor** [comment]
@SimonSapin: sorry, I should have also linked to the comments above. I think the status is that, modulo outstanding concerns from some users about the naming, these methods are probably suitable for stabilisation.

> Is this still the case or has some consensus been found since?

The current names [were suggested](https://github.com/rust-lang/rfcs/pull/2757#issuecomment-564683344) by @dtolnay. I'm personally happy with the existing names, but there have been some complaints: https://github.com/rust-lang/rust/issues/64260#issuecomment-578509401, https://github.com/rust-lang/rust/issues/64260#issuecomment-579997086. However, I think considering how much previous discussion we had on these names, without coming to a clear consensus, I think it's probably worth stabilising with the existing names, so these methods can be used.

Perhaps starting an FCP would be the best way forward with these methods now: if any libs team members decide there are better alternatives, these could be raised during the FCP — otherwise, it would be good to be able to use these on stable.

**Comment by SimonSapin** [comment]
As a reminder, this issue currently tracks:

```rust
impl bool {
    pub fn then_some<T>(self, t: T) -> Option<T> {…}
    pub fn then<T, F: FnOnce() -> T>(self, f: F) -> Option<T> {…}
}
```

I can live with this, and more time likely won’t bring new information or ideas at this point.

@rfcbot fcp merge

**Comment by Boscop** [comment]
IMO if we only have these 2 methods, there should **definitely** be another method to turn a bool into an `Option<()>` (shorter than `b.then_some(())`).
I.E.:
```rust
impl bool {
    pub fn opt(self) -> Option<()> {...}
}
```
This corresponds to [`Boolinator::as_option`](https://docs.rs/boolinator/2.4.0/boolinator/trait.Boolinator.html#tymethod.as_option).
If we only have `then_some` (aka `Boolinator::as_some`) and `then` (aka `Boolinator::as_some_from`), 
**then** boolinator's `b.and_option(opt)` can be expressed as `opt.filter(|_| b)` but we have no equivalent to `Boolinator::and_option_from`:

```rust
fn and_option_from<T, F: FnOnce() -> Option<T>>(self, f: F) -> Option<T> 
```
If we had the `opt` method above, we could express this as `b.opt().and_then(|_| f())`. 
Without this `opt` method, we'd have to write `b.then_some(()).and_then(|_| f())` which is much less readable, so it'd be less ergonomic than `boolinator` :/

**Comment by dtolnay** [comment]
@Boscop, be aware that we are not aiming to make boolinator obsolete. It isn't a goal of this issue or in general to absorb everything in boolinator into the standard library, or to expose at least as concise standard library equivalents of everything that boolinator exposes. In general external libraries get to cater to quite different tradeoffs than a standard library can (on many dimensions: nicheness, quirkiness, etc) so those people who have been happy with boolinator should go on using boolinator.

**Comment by voidc** [comment]
How about also adding a method with `bool` and `T` switched.

    trait TakeIf {
        fn take_if(self, b: bool) -> Option<Self>;
    }

    impl<T> TakeIf for T {
        fn take_if(self, b: bool) -> Option<Self> {
            b::then_some(self)
        }
    }

This would improve the ergonomics of using this feature in method chains.

**Comment by SimonSapin** [comment]
@voidc The bar for adding a method to *every single type ever* should be very high. This is rather niche functionality, so I feel it doesn’t deserve to be in the prelude. If importing a trait is required, this might negate the convenience you would find in the switched order.

**Comment by withoutboats** [comment]
@rfcbot concern name

I do feel like users (including myself) may have trouble remembering these APIs because in contrast to and/and_then and or/or_else, the long form is the non-closure form. However, I'm not clear what the alternative would be.

Assuming that we are adding these two methods only, can anyone come up with a pair of names that seem fitting and appropriate for boolean methods, where the one that takes `T` by value is a substring of the one that takes a closure?

**Comment by djc** [comment]
Summarizing previous discussion from the [RFC PR](https://github.com/rust-lang/rfcs/pull/2757). They were originally called `then()` and `then_with()`, then `to_option()` and `to_option_with()` (both of these would satisfy your concern). @dtolnay proposed the current names in https://github.com/rust-lang/rfcs/pull/2757#issuecomment-544277420 and elaborated on the choice in https://github.com/rust-lang/rfcs/pull/2757#issuecomment-544581104.

Closely related combinators as summarized by @troiganto in https://github.com/rust-lang/rfcs/pull/2757#issuecomment-529151770:

```rust
x.ok();  // Result<T, E> → Option<T>
x.err(); // Result<T, E> → Option<E>

x.ok_or(y);            // Option<T> → Result<T, E>
x.ok_or_else(closure); // Option<T> → Result<T, E>

x.and(y); x.and_then(closure); // Result<T, E> → Result<U, E>
x.or(y); x.or_else(closure);   // Result<T, E> → Result<T, F>

x.and(y); x.and_then(closure); // Option<T> → Option<U>
x.or(y); x.or_else(closure);   // Option<T> → Option<T>
```

**Comment by Boscop** [comment]
@withoutboats The names from `boolinator` fulfill your criteria as well: 
`as_some` and `as_some_from`. Also adhering to the convention that `*_from` will create it from a given closure.

But I think the names `then`/`then_with` are even better (and there won't be confusion since we won't have something like `and_then` for bool).

**Comment by withoutboats** [comment]
Thanks for linking to the previous discussion @djc. Lookin at the code I found it actually fairly intuitive in practice and I'm not as concerned about remembering which is correct. I think in practice, the notion of "then <take some action>" and "then_some <value>" is good to grok.

But I don't really see a strong argument made in the old thread between then/then_with and then_some/then. I strongly agree with @dtolnay's argument about names like to_option, but is there a reason to prefer then_some/then over then/then_with?

**Comment by jplatte** [comment]
Personally, I like the shorter variant being the one that takes a closure, as when I tried out the feature, I never (or almost never) used the variant that takes a value.

**Comment by withoutboats** [comment]
@jplatte Yea, that seems like what I would expect as well. I think of this as replacing this common and annoying pattern:

```rust
if foobar {
   // code block evaluationg to Option<T>
} else {
   None
}
```

Rarely is the `if` case just literally `Some(thing)`, it almost always involves running some sort of code, no?

I wonder what people think of a third option: only adding `then` and expecting people to write `foobar.then(|| Some(thing))` instead of `foobar.then_some(thing)`?

**Comment by varkor** [comment]
> Rarely is the if case just literally `Some(thing)`, it almost always involves running some sort of code, no?

It my experience, it's frequently just `Some(thing)`. You can see this [in rustc](https://github.com/rust-lang/rust/search?q=then_some&unscoped_q=then_some) for instance, although that's certainly an under-approximation. I think both methods are convenient in practice.

**Comment by djc** [comment]
But we can do `then(|| thing)` without the `Some()` right? At least the RFC seems to mention that the closure parameter `F` would just return `T` directly.

I do think that the closure form would be quite a bit more likely, and the syntactic overhead of `|| ` here seems very limited. It does cause some asymmetry with other pairs of methods. What is the history of `Option::unwrap_or()` vs `Option::unwrap_or_else()`? If the latter had been there first, would we still have the shorter version?

Also, is there a performance impact of the extra method call, or will LLVM know to inline it basically always in the common case?

**Comment by dtolnay** [comment]
I agree with https://github.com/rust-lang/rust/issues/64260#issuecomment-609033148 that both forms are frequently applicable in my code. I would end up using `then_some(thing)` wherever applicable because it reads off more clearly than `then(|| thing)` -- it means `if ... then Some(thing)`. I think `then` is better fit to blocks of code where it means "if ... then do the following".

```rust
is_true.then(|| {
    ...
    ...
})
```

**Comment by Boscop** [comment]
So we cannot call `then` if the closure returns an Option. What's the suggested way to shorten this then?
```rust
if cond {
    f()
} else {
    None
} 
```

**Comment by dtolnay** [comment]
`cond.then(f).flatten()`

**Comment by U007D** [comment]
I find `then` and `then_some` difficult to remember, because they do not follow the patterns of existing similar combinators, as @djc illustrated above.  Further, I find them to be named backward--I tend to reach for the lazily-evaluated version as opposed to the eager one.  Everywhere else in `std`, the shorter name is eager, and the name with the suffix (*_then, *_with, etc.) is lazy.  Unfortunately the `bool` ones are currently reversed.

I prefer names that require low cognitive load e.g. `to_option/to_option_with`.  Failing that, providing the extended suffix name to the lazy version e.g. `and/and_then` (or `then/then_with` if we must) would at least be consistent with the rest of `std`.

Thanks for reading; I know this can be draining with everyone having a strong opinion.

**Comment by dtolnay** [comment]
I am going to lock this for a while because it doesn't seem like new arguments or new ideas are still being introduced.

There's a good chance we'll remove these methods; as described in https://github.com/rust-lang/rfcs/pull/2757#issuecomment-552343795, they were landed unstable to find out whether we might settle on sufficiently clear names for the behavior based on actual use, which I guess hasn't happened.

**Comment by varkor** [comment]
> I am going to lock this for a while because it doesn't seem like new arguments or new ideas are still being introduced.

While I appreciate that comments reiterating previous points are not helpful, locking the thread is a very effective way to ensure that no new arguments or ideas can be presented.

> There's a good chance we'll remove these methods; as described in rust-lang/rfcs#2757 (comment), they were landed unstable to find out whether we might settle on sufficiently clear names

While it seems increasingly clear that there do not exist names that makes everyone happy, I think it's also plain to see that these methods are desirable to have in the standard library. As a case in point, many of the existing combinators have entirely unenlightening names, but are nevertheless useful; though I still forget some of the combinators, I am happier to have to check every so often, than to have to rely on a library or my own implementation for such a common idiom. The same is true of the methods here.

I would imagine that even those with objections to the current names would still prefer to have the methods as-is than not at all (though this would be a little unfair to claim when they do not have the chance to respond).

**Comment by LukasKalbertodt** [comment]
I agree with @varkor and don't think locking this thread really helps the discussion. There have been threads way worse than this in Rust's history ;-) 
I will just go ahead and unlock this thread again. If the majority of the libs or moderation team is in favor of locking, just undo that.

---

On the topic: I truly appreciate you (@dtolnay) expressing the often unpopular opinion of removing a feature instead of stabilizing. And I even agree that a feature which is mainly added for convenience and brevity of code shouldn't be added if it makes code notably less readable. However, in this case, I think we gain more by having the method than we lose by the non-perfect name. 

[In your example](https://github.com/rust-lang/rfcs/pull/2757#issuecomment-544277420), we replace a few perfectly clear lines with one line that the reader has to briefly think about. Removing lines is not just about less typing, but also about removing boilerplate code that does not carry information. Lots of boilerplate code also makes functions hard to understand and exhausting to read. So in my book, using the new method is not worse than the old code.

**Comment by dtolnay** [comment]
Locking is valuable as a triage tool -- if a feature is relatively low value but looks to have disproportionately high cost to reach a consensus on, there are going to be more productive uses of time. Higher value features than this get to have proportionately more intense discussion.

I guess I would just remind participants that repeating previous comments without adding new information is not useful. You would need to acknowledge (ideally link) and address some of the counterpoints that have already been made against your position in a way that hasn't already been discussed.

**Comment by Boscop** [comment]
> I would imagine that even those with objections to the current names would still prefer to have the methods as-is than not at all (though this would be a little unfair to claim when they do not have the chance to respond).

I totally agree. I'd rather have these methods as-is in std than not have them. In fact, after re-reading this thread and using these methods in my own code, I've now come around to liking these names and got used to them. So yeah, please add them :)

**Comment by withoutboats** [comment]
> There's a good chance we'll remove these methods; as described in rust-lang/rfcs#2757 (comment), they were landed unstable to find out whether we might settle on sufficiently clear names for the behavior based on actual use, which I guess hasn't happened.

I wouldn't want that to be the outcome.

I actually think this is a relatively high impact change - I suspect there's a good chance that using this API will become a very common pattern, part of the essential toolbox of methods all users are expected to know. That's why I think quibbling over these very small differences is worthwhile.

I see three outcomes that are, in my mind, "on the table:"

1. Stabilize as is: `then_some`/`then`
2. Change the names to `then`/`then_with`
3. Only stabilize `then`

3 is forward compatible with 1, otherwise these are incompatible. So we need to make a decision.

I agree with @Boscop's comment that seeing these in actual code made option 1 much more appealing than hearing it in the abstract - both names are rather clear about what they do.

I feel uncertain that `then_some` pulls it weight, though. Here are the downsides, to me, of adding `then_some`:

1. It is another method in this core toolbox of methods that everyone is expected to learn. We've very rarely added methods to this toolbox since 1.0, and among those we have added, many I think we should've been more cautious about (like `Option::xor`).
2. It breaks the pairing of short:value/long:closure that the pair methods on Option and Result have. I think that's been a big boost for making learning these methods easier, because they pair off in this consistent way.

(In fact, `then` alone would boost the effect of these pairings by tightly linking the idea of then to the idea of taking a closure in the std APIs.)

I agree that ìf_foo.then_some(x)` is a bit clearer than ìf_foo.then(|| x)`, but I'm not completely convinced that outweighs these downsides. What do you think, @dtolnay, about only stabilizing `then` and revisiting later if it feels like the pattern `then_some` supports needs this further clarity enough to add this method?

**Comment by dtolnay** [comment]
I would be on board with stabilizing `then` and deferring until later whether to remove or stabilize `then_some`.

**Comment by pnkfelix** [comment]
I already accidentally wrote a note on the related RFC (https://github.com/rust-lang/rfcs/pull/2757#issuecomment-614126309); I just wanted to chime in saying that I have independently encountered the scenario outlined by @U007D [up above](https://github.com/rust-lang/rust/issues/64260#issuecomment-578509401), and that indicates to me that we should consider making `bool::then` take an `FnOnce() -> Option<T>` instead of an `FnOnce() -> T`.

In my aforementioned [note](https://github.com/rust-lang/rfcs/pull/2757#issuecomment-614126309), I suggested that both `bool::then` and `bool::then_some` should take closures (and the difference is just that `then_some` takes an `FnOnce() -> T`).

(I did see the [response](https://github.com/rust-lang/rust/issues/64260#issuecomment-578517822) pointing out that one can use `Option::flatten` to get around the issue of passing an `FnOnce -> Option<T>` into the current `bool::then`. I haven't made up my mind yet about whether that would be a reasonable solution in most cases; for the one that I was facing most recently, it made me abandon using these combinators entirely.)

**Comment by varkor** [comment]
> that indicates to me that we should consider making bool::then take an FnOnce() -> Option<T> instead of an FnOnce() -> T.

It indicates to me that, if this is such a common scenario, we ought to have a third method to avoid `.flatten()`. However, part of the motivation for these combinators is to avoid having to write `Some` everywhere, which would no longer be the case if the combinators return `Option<T>`.

**Comment by Boscop** [comment]
> It indicates to me that, if this is such a common scenario, we ought to have a third method to avoid .flatten().

I think it makes sense to add this third method, and it could be called `and_then`.

**Comment by traviscross** [comment]
Speaking as someone who trips over the absence of this feature somewhat frequently, I'd be happy to have this under any reasonably concise name.  But I agree with @withoutboats that this will be an important idiom in the language, so we should get it right.

Also agreeing with @withoutboats, I could live a happy life without `.then_some`.  Writing the closure is fine as our closure syntax is concise.  (Even better, not having the method means clippy won't warn about using `.then(|| x)` instead of `.then_some(x)`).

However, as @varkor and @Boscop have pointed out, it would be nice to have a method that does not implicitly wrap with `Some`.  (I would run up against this in some code I'm working on today.)  So if we want to add two methods and check our happiness with the consistency of those names, it seems those are the two that we should consider.

One (perhaps radical) option would be to add just one method, `.then`, and have it not do implicit wrapping.  Personally, I could live with writing the explicit `Some`.  Saying  `is_foo.then(|| Some(42))` is still infinitely better than today's idiomatic code.  (Naming this `.and_then` would also be fine.)

That also solves the worst problem with using the name `.then` for a method that wraps its result with Some, which is that our most prominent other "then-like" method, `.and_then`, doesn't implicitly wrap.

We could always decide later, on the basis of experience, to add a method that does implicit Some-wrapping.  (I'd suggest `.map` as the name of that method.)

**Comment by traviscross** [comment]
For concreteness, here's a proposal that maximizes parsimony with the rest of Rust:

```rust
fn and_then<U, F: FnOnce() -> Option<U>>(self, f: F) -> Option<U> {
    if self { f() } else { None }
}
fn map<U, F: FnOnce() -> U>(self, f: F) -> Option<U> {
    self.and_then(|| Some(f()))
}
```

Play with them here: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=259b06d354e96e9e29fb883e25c98215

The way that I justify these names to myself is that they first implicitly lift the bool into the Option monad.  The only other reasonable behavior for a `.map` method would be to map `bool -> bool`, and it just doesn't seem likely that we'd ever add that (cf. `is_foo.and_then(|| None).is_some()`).

**Comment by traviscross** [comment]
Thinking about this further, if there were a method named `then`, [here's](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=685572237a921b4684c4bd2603ed73b1) what I would expect it to do:

```rust
fn then<U, F: FnOnce(Self) -> U>(self, f: F) -> U {
    f(self)
}
```

E.g. `is_foo.then(|x| if x { Ok(42) } else { Err(()) })`.  I.e., it would allow arbitrary logic within a method chain, similar to [expression-oriented keywords](https://boats.gitlab.io/blog/post/await-decision/) (e.g. `is_foo.if {...} else {...}`).  Intuitively, `and_then` feeds forward only the true/present/success value, so `then` should feed forward all values.

Note that I'm not proposing this method here for inclusion.  It's just what I would expect a `then` method on a Rust type to do given other method names in the language.  It's one more reason why `and_then` is probably the right method name.

**Comment by zakarumych** [comment]
For the case mentioned above
```
impl Foo {
    fn new() -> Option<Self> {
        step_1_check().and_then(|| step_2_check().map(|| Self { /* ... */ }))
    }
}
```

wouldn't it look nice to write this instead

```
impl Foo {
    fn new() -> Option<Self> {
        step_1_check()?;
        step_2_check()?;
        Self { /* ... */ }
    }
}
```

Which would be possible by implementing `std::ops::Try for bool`

**Comment by jdahlstrom** [comment]
If it's too difficult to come up with proper names for the combinators, couldn't we just have a `bool -> Option<()>` conversion function as a first step at least? Those two types are famously isomorphic (and even have identical representation), so something like `bool::as_option` shouldn't be too far-fetched. With just one extra method call you then get access to everything `Option` has to offer.

**Comment by workingjubilee** [comment]
I was hand-rolling this pattern in code yesterday.

I don't believe bool should necessarily have a 1 to 1 parity in names and methods with Option or Result, and I think that if `map` appears on things that aren't already "wrapping" types (iterators or options), it could just be confusing. But I agree with `and_then` existing and carrying an equivalent signature to Option's version, partly because it would work well with  `filter_map` (that was my original intent, in fact).

**Comment by ZaneHannanAU** [comment]
Found out about this today.

If possible, could we get an `or_else` and an `or_else_some`?

My current option is a bit... awkward:

```rust
  let rustls_cfg = (!no_tls)
    .then(|| -> Result<ServerConfig, Error> {
      let mut config = ServerConfig::new(NoClientAuth::new());
      let mut cert_file = BufReader::new(File::open(cert_file)?);
      let mut key_file = BufReader::new(File::open(key_file)?);
      let cert_chain = certs(&mut cert_file)?;
      let mut keys = rsa_private_keys(&mut key_file)?;
      config.set_single_cert(cert_chain, keys.remove(0))?;
      Ok(config)
    })
    .transpose()?;
 ```

But that's also because I wanted to support having no-tls on for testing purposes (as well as because it's difficult to get the keys without having some keys anyway...)

**Comment by sanbox-irl** [comment]
What about just...`some` and `then_some`?

`some` is the non-closure version, `then_some` is the closure version.

I feel like it seems pretty clear since it literally has the word `some` in it, so you get nice code like:

```rs
let enemy = is_bad_guy.some(thing);

let find_other_enemy = is_bad_guy.then_some(|| find_thing());
```

As others have said, it feels like this would be a good thing to have in some capacity at all, even if it's just an easy conversion

**Comment by hniksic** [comment]
> `bool::as_option` shouldn't be too far-fetched. With one method call you get access to everything Option has to offer.

This is an excellent proposal. `bool::and_then(...)` and similar feel wrong, first because `bool` is not an option and these methods conflate the two steps (converting to `Option` and doing something withit). The second reason it feels wrong is that it requires the **second** method, `then_some`, which starts to feel like overkill.

But with `bool::as_option()` that returns `Option<()>`, we get:

* a *single* method on `bool` with minimum pollution and maximum clarity;
* `and_then` and `then_some` are automatically expressed with `Option`'s own `and_then`, `map`, and others.

`Option` has a well-deserved abundance of methods for transformation. Let's use not reinvent them in `bool`, let's just provide an easy and obvious way to access them.

**Comment by djc** [comment]
Going to try to summarize updates since my [last summary](https://github.com/rust-lang/rust/issues/64260#issuecomment-608407979)) from 4 months ago:

@withoutboats initially canceled FCP with the following concern:

> Assuming that we are adding these two methods only, can anyone come up with a pair of names that seem fitting and appropriate for boolean methods, where the one that takes T by value is a substring of the one that takes a closure?

1. The initial RFC proposal was actually:
  * `fn then<T>(self, t: T) -> Option<T>`
  * `fn then_with<T, F: FnOnce() -> T>(self, f: F) -> Option<T>`
2. Before the RFC was accepted, they were briefly called:
  * `fn to_option<T>(self, t: T) -> Option<T>`
  * `fn to_option_with<T, F: FnOnce() -> T>(self, f: F) -> Option<T>`
3. **In the RFC we have**
  * `fn then_some<T>(self, t: T) -> Option<T>`
  * `fn then<T, F: FnOnce() -> T>(self, f: F) -> Option<T>`
4. As mentioned by @Boscop, the boolinator crate provides
  * `fn as_option(self) -> Option<()>`
  * `fn and_option_from<T, F: FnOnce() -> Option<T>>(self, f: F) -> Option<T>`
5. One option would be to only stabilize one ([proposed](https://github.com/rust-lang/rust/issues/64260#issuecomment-610965009) by @withoutboats -- @dtolnay [liked](https://github.com/rust-lang/rust/issues/64260#issuecomment-611035438) this option):
  * `fn then<T, F: FnOnce() -> T>(self, f: F) -> Option<T>`
6. @pnkfelix [suggested](https://github.com/rust-lang/rust/issues/64260#issuecomment-614134947) that the non-`Some` version should also take a closure:
  * `fn then_some<T, F: FnOnce() -> Option<T>>(self, f: F) -> Option<T>`
  * `fn then<T, F: FnOnce() -> T>(self, f: F) -> Option<T>`
7. @varkor then [commented](https://github.com/rust-lang/rust/issues/64260#issuecomment-6141527680) that one of their goals was to be able to do without explicit `Some`, and maybe a third method should then be added, @Boscop called it `fn and_then<T>(self, T) -> Option<T>`
8. @sanbox-irl [suggested](https://github.com/rust-lang/rust/issues/64260#issuecomment-642886982):
  * `fn some<T>(self, t: T) -> Option<T>`
  * `fn then_some<T, F: FnOnce() -> T>(self, f: F) -> Option<T>`
9. @traviscross [looked at](https://github.com/rust-lang/rust/issues/64260#issuecomment-618681388) other combinators and suggested this might have the best similarity with existing methods (echoing earlier [comments](https://github.com/rust-lang/rust/issues/64260#issuecomment-578509401) by @U007D):
  * `fn and_then<U, F: FnOnce() -> Option<U>>(self, f: F) -> Option<U>`
  * `fn map<U, F: FnOnce() -> U>(self, f: F) -> Option<U>`
10. @workingjubilee [suggested](https://github.com/workingjubilee) it should work well with `filter_map()`
11. @jdahlstrom [suggested](https://github.com/jdahlstrom) we should have `impl From<bool> for Option<()>`, which seemed fairly popular

Phew, hope I represented everyone correctly. If not, let me know and I'll update this post.

I think there's some Stroustrup's Rule aspect to it, too. If this becomes a common idiom in the language (as withoutboats [suggested](https://github.com/rust-lang/rust/issues/64260#issuecomment-610965009)), it makes sense to have multiple methods to cover all the bases. From that perspective, it makes some sense to start by stabilizing a single method.

In that case, the original proposal for this was `fn then<T, F: FnOnce() -> T>(self, f: F) -> Option<T>`, which seems sensible enough. The lexical overhead of closure wrapping seems limited (just `|| `), it accomodates for laziness in the case of expensive `T` and doesn't require spelling out `Some`.

Comparatively, something like `as_option()` is also a minimal proposal, adding only one method. On the other hand,  it will require an extra combinator at every callsite, so it's more wordy (in particular, in rustfmtted long combinator chains it will require an extra line). I do think adding this as a `From` impl is a neat idea, but I think using it would require turbofishing when used in the middle of a combinator chain.

It is true that `then()` and `then_some()` are not the most consistent with existing methods from `Option` and `Result` (neither of those have a `then()` method, although `FutureExt` does). However, `map()` doesn't seem like a great alternative to me in terms of naming because unlike other uses of `map()`, this doesn't keep the same "outer" type.

I hope we can move this forward!

**Comment by jdahlstrom** [comment]
> Going to try to summarize updates since my [last summary](https://github.com/rust-lang/rust/issues/64260#issuecomment-608407979) from 4 months ago:
>
>     2. @jdahlstrom [suggested](https://github.com/jdahlstrom) we should have `impl From<bool> for Option<()>`, which seemed fairly popular

To clarify, I only meant to propose adding *some* function of type `bool -> Option<()>`, not a `From` impl specifically. (My bikeshed name was `bool::as_option`, which is what boolinator also has so there's some precedent). But a `From` impl would be fine as well, assuming turbofish isn't needed too often.

**Comment by runiq** [comment]
@djc Would you consider adding @hniksic's [proposal](https://github.com/rust-lang/rust/issues/64260#issuecomment-681726105) too? It sounds like the most straightforward to me.

**Comment by djc** [comment]
@jdahlstrom / @runiq: done (consider hiding your comments as resolved to make this trend easier to follow).

**Comment by Alfriadox** [comment]
What about a function `on_true` that takes a closure and returns an option as I think has already been discussed here.

**Comment by hniksic** [comment]
@djc Please merge my and @jdahlstrom's proposal, as they are the same. My comment just argued for it more elaborately because I thought it was under-appreciated and possibly went under the radar.

For example:

11. @jdahlstrom proposed and @hniksic argued for a simple `bool::as_option() -> Option<()>` conversion, which provides a single method and leaves further transformations to methods already existing on `Option`.

**Comment by glaebhoerl** [comment]
Just a very small note (based on Ctrl-F for various words it doesn't *seem* to have been specifically mentioned before) that one of the benefits of consistent naming for methods on `bool` resp. `Option` would be the ability to "guess" their existence, without having to look it up in the docs -- "I know this method I need exists for `Option`, hmm, *start typing*, ah, it exists for `bool` too, cool!". 

(As usual with this kind of thing, the tradeoff is that the names might make less sense when considered only in the context of `bool` itself. I'm not endorsing either half of this tradeoff as being the more important one, just pointing it out.)

**Comment by Alfriadox** [comment]
I think that since `bool`s are language primitives, it's acceptable for them to break with the naming style on `Option` and `Result`.
In general forcing everything to match the convention only makes sense when they are used for the same purpose. `Option` and `Result` are both generic and store something. Methods like `then` or `map` indicate to me the operating on a value stored inside the struct instance. `bool`s do not store values, and therefore, in my opinion, do not need to follow convention. That's why the method (or methods) should be named (`on_true`, `do_on_true`) or if we were to match the functions in the `boolinator` crate, (`as_option`, `as_option_from`). Arguably the boolinator names are better, since they have the word option in them (and for people programming in an IDE, typing option would bring up auto complete.) but I personally like `on_true` because it feels like it better indicates the Boolean-first nature of the operation.

**Comment by workingjubilee** [comment]
I acknowledge that the `map`/`and_then` parity would be great for discoverability, but I am skeptical about after. Yes, people do think about code in terms of raw literal chars and what raw chars we can put in, in other words, typographically, to get certain results. So it's obvious it would help that. But our ability to quickly **exploit** discoveries actually depends on the consistency of less immediate things than just knowing the name: namely, what the name represents.

In Rust, `map` and `and_then` (AKA flatMap) depend, conceptually, on having containing values. This is not just an idle characteristic of them: it defines **every** function signature for them! They ALL accept an Fn that accepts T and returns U, to get the result of a wrapped value. That means these closures have an argument. Not zero, but one.

An experienced Rustacean may personally already have learned this, and be able to implement the variation in their mind, however, this doesn't lend well to learning because **many** people come from other languages that lean hard on if/else boolean logic, not having `match`-like constructs, and they write highly `bool`-oriented code in Rust, even to the point of reinventing `match`. So we have a question: isn't it likely that they will learn, but then find it conflicts with 80% of the other usages? Isn't that injecting a lot of confusion into `std`?

In addition, even for an experienced Rustacean, and in fact even moreso per day of experience one has had, `map` and `and_then` having such levels of consistency allows one to gain the ability to write such code without even thinking about it, it almost lives in the fingertips. That's fluency. I think it will cause quite a few stumbles in actuality, regardless of experience, and it is offering apparent, not actual, harmony with existing APIs. Sure, eventually everyone will overcome this weird bump in `std`, but it's one that doesn't need to be introduced.

As for the rest, `as_option` probably breaks with the existing [API naming guidelines](https://rust-lang.github.io/api-guidelines/naming.html) on conversions, `as` is usually for borrowed types. `to_option` would be preferred, I think. I like [sanbox-irl's `some`](https://github.com/rust-lang/rust/issues/64260#issuecomment-642886982), but I also like `then`. We probably peaked with `then`, really. It's a natural follow-on from a `true` statement in many PLs.

**Comment by ZaneHannanAU** [comment]
I think on_true implies that it'd always be the default value on false, ie

```rust
impl<T: Default> bool {
  fn on_true(self, ex: FnOnce() -> T) -> T {
    if self {
      ex()
    } else {
      T::default()
    }
  }
}
```

which is fine, but not great. My personal preference now is towards `and_then<T> -> Option<T>`, `and_then_else<T, E> -> Result<T, E>`, or possibly `unwrap_true<T> -> T`. Simply put:

```rust
trait Sealed {}
pub trait BoolExt: Copy + Sealed {
  fn and_then<F, T>(self, if_true: F) -> Option<T> where F: FnOnce() -> T;
  fn and_then_else<FT, T, FE, E>(self, if_true: FT, if_false: FE) -> Result<T, E> where FT: FnOnce() -> T, FE: FnOnce() -> E;
}
impl Sealed for bool {}
impl BoolExt for bool {
  fn and_then<F, T>(self, if_true: F) -> Option<T> where F: FnOnce() -> T {
    if self {
      Some(if_true())
    } else {
      None
    }
  }
  fn and_then_else<FT, T, FE, E>(self, if_true: FT, if_false: FE) -> Result<T, E> where FT: FnOnce() -> T, FE: FnOnce() -> E {
    if self {
      Ok(if_true())
    } else {
      Err(if_false())
    }
  }
}
```

**Comment by workingjubilee** [comment]
So, while I said that I don't think `map` is a good idea, I also no longer think that `then` or `then_with` should be used because I noticed [Ordering](https://doc.rust-lang.org/std/cmp/enum.Ordering.html) uses both of those names for a very different purpose. I now am inclined to say that `some` and `then_some` are the best versions I've seen to date.

**Comment by sanbox-irl** [comment]
Hello!

I previously have argued for:
```
fn some<T>(self, t: T) -> Option<T>
fn then_some<T, F: FnOnce() -> T>(self, f: F) -> Option<T>
```

I would like to discuss `some` as the option for taking a `value` here.

1. We should follow the shape of `Option<T>`.

Ideally, the entire `std` followed the same pattern of `value/lazy` functions for their combinators, especially regarding the rather unique English word "then", but that isn't the case (and perhaps is a good argument that such a thing *wouldn't* be unqiue -- I'm very convinced by @workingjubilee 's argument that English meaning fades relatively quickly).

However, we don't really need to follow `std` -- we should just be following `Option`, because this is a group of functions designed to work *with* option. `Ordering`, `Result`, and other std constructs are important reference points, but I think similarity to `Option` is the most important, because users will be transitioning mentally "into" the Option part of their brain. Having those APIs have a similar pattern of usage seems very good to me -- on the other hand, the utility of being similar to `Result` is more "ambient", which has value, but I think not as much. As a metaphor, I'd say a given website is benefitted if it is similar to other websites users will encounter, but a *page* on a website *really really* needs to be similar to the other pages on a website. In this sense, even though this API will be in the `bool` area of the `std`, I'd really encourage us to think of it as being part of `Option`.

2. `Option` uses a longer version of a shorter name in general to express combinators that are lazily evaluated.

@withoutboats referenced this pattern, and I think we ought to map after this shape.
here's option's API which has this duality (and sometimes a *triality* is that word i don't think it is):

`and/and_then`  (uses `then`)
`get_or_insert/get_or_insert_with` (uses `with`)
`map/map_or/map_or_else` (uses `else`, but this is the exception, as `map` also takes a function, and `map_or` doesn't, which might be why `map_or` confused the hell out of me)
`ok_or/ok_or_else` (uses `else`)
`or/or_else` (uses `else`)
`unwrap_or/unwrap_or_else` (uses `else`) 

So based on this point and on the excellent summary by @djc above, the following which have been suggested would follow this pattern in `Option`:

`then/then_with` (the original proposal)
`as_some/as_some_from` (boolinator)
`some/then_some`

I'd like to talk specifically about the proposal to stabilize only `then`, as `fn then<T, F: FnOnce() -> T>(self, f: F) -> Option<T>`, and stabilize a by value variant at a later time. I would argue against doing that because `then` prevents us from following the similar `Option` patterns, which would be a shame I think.

3. The "hot path" for converting a `bool` -> `Option<T>` is for a `true` -> `Some(T)`, not `false` -> `None`

The most common way that I use this function, in the Boolinator crate, is in the following pattern:

```
struct Foo {
    success: bool,
    value: i32, // for example...
}

fn main() {
    let input = vec![
        Foo {
            success: false,
            value: 10,
        },
        Foo {
            success: true,
            value: 5,
        },
        Foo {
            success: true,
            value: 10,
        },
    ];

    let big_value: i32 = input
        .into_iter()
        .filter_map(|v| v.success.as_some(v.value))
        .max()
        .expect("look it's an example");
}
```

There are other options as well, but ultimately, my point here is that the conversion from `bool` to `Option` is *really* about converting a `true` to a `Some(T)`, not about converting a `false` to a `None`, because the idea is really to filter out times when some `bool` is `false` to better access the `T`.

We see this not only on a structural level with how boolinator is used, but also how people name their `bools` -- in my experience, `bools` are named where *true is the probable state*. For example `is_success`, `winning`, `allocated`, etc. I more rarely see negatively phrased boolean names, like `is_poisoned`, `is_closed`, etc (though they *absolutely* exist of course -- these are all real examples from the std).

To attempt to speak like I know what I'm talking about (which is this entire post), that occurrence is what allows us to generalize that `true` => Some, and `false` => `None`, which is the purpose of the RFC. Therefore, `is_success.some(10)`, when `is_success` is false, is not the "mental" path that the code writer is writing for -- they're thinking in terms of positive success being converted to positive `Some`.

To that end, I think this function should include the word `some`, not the word `Option`.

Therefore, out of the options above, I think the best remaining options are:
`as_some/as_some_from`
`some/then_some`

and descriptiveness between equal between them, length matters, because as has been said, this function will likely become a stable part of a Rust programmer's toolkit, and therefore, I prefer `some/then_some`. Wow, it sounds so incredibly arrogant to say that about my idea -- I apologize if I seem dismissive of anyone else's ideas; I'm presenting this argument a bit more strongly than I feel since I think we all know that a decision is a bit overdue, so I'm attempting to not allow the very good be the enemy of the perfect.

I think the biggest weakness of `some`, however, is in the following case:

```
fn main() {
    let nope = false;
    let option = nope.some(10);
}
```

The confusion there that a function named `some` actually is producing a `None` is not insignificant, but I think is overwhelmed by the benefits.

I have largely been discussing `some` rather than `then_some` or the other lazy variants because my original proposal was for `then_some`, but now that I've looked at it more deeply, I am most in favor of `some/some_with` actually, but I don't want to add too much confusion by clouding the waters.

As most have said, I agree that the lazily evaluated one will probably see more use, and can be a stooge for the one by value, while the value cannot replicate the lazy option, but I think we have to still name the `value` one first, because we will build the lazy function name on top of the `value` one. I propose, therefore, that we name it simply `some`.

I believe `some` is the best options because it links the API directly to `Option` by name, it maps better to how `Option` works, and it better represents how people write code with a preference towards `true` values.

Apologies for writing so much!

What does everyone else think?

**Comment by Alfriadox** [comment]
Good post overall. I worry that making the function name `some` may generate some confusion when a false value constructs a none. I also believe that we should be putting more emphasis on the boolean side than the option side. Your deep dive into the suffixes used for lazy versions of functions was good though. I think it helped me realize that I view the lazy version of this function as more separate from the value version than it is for other types (Option, Result, etc). I also doing have the notion that booleans are generally true positive. My experience does not lead me to believe that booleans generally lean one way or another, or have a conventional meaningfulness to them. Even if they did, I think it would be an overly specific case to design for.

With that all said, I think our best options going forward are `on_true`/`if_true`. These both reflect the booleans use as a tool for control flow (which is ultimately what it is in the pattern being replaced).

**Comment by sanbox-irl** [comment]
I agree that the weakness of `some/some_with/some_then` is the oddness of false.some() -> None, and that might be a big enough weakness to others to reject it, but personally, as I said, I think of `bool` -> Option as really converting a `true` to a `Some<T>`, and a `false` -> `None` being a thing which also happens, but not the purpose of the operation. I think that's what I'm trying to get at that there -- we convert `bool` -> `Option` *because* of the `true` case, not because of the false case -- `None` contributes exactly the same amount of information as `false` does (`Option<()>` and `bool` are isomorphic), but `Some<T>` is a more useful construct than just a `bool`, which is why we do the operation.

**Comment by ErichDonGubler** [comment]
@sanbox-irl: My first thought about `bool::some` looking strange at the call site is that that might be alleviated by infusing some of the other semantics suggested into it, i.e., `and_some`, `and_then_some`:

```rust
rust_is_awesome.and_then_some(|| "woot")
failed.and_some(success_value)
```

**Comment by sanbox-irl** [comment]
I much like the above and option. That neatly aligns with how Option's and* methods return another Option. 

@Alfriadox im not as much of a fan of that naming scheme. Although it captures the structural change, I think it might oversell because it doesn't reference Option very well. It actually sounds like a function that would take an Fn -> () to me.

**Comment by Alfriadox** [comment]
Honestly, if we cannot agree on anything else, I think that I'm alright with the originally proposed function names (`then_some`/`then`). I prefer those over the `and*` naming, as its too close to the logical and (`&&`) operation for me.

**Comment by thomaseizinger** [comment]
Not sure if it has been pointed out already but there is an inconsistency within the std-lib on the conversions between `Result` and `Option` which might influence how we want to name the conversion functions between `bool` and `Option`.

In particular:

Going from `Result` to `Option` is called [`.ok`](https://doc.rust-lang.org/std/result/enum.Result.html#method.ok), being named after the variant of the `Result` you have. In other words, "If it is `ok`, give me a some" and "If it is `err`, give me a some".

On the other hand, the same functionality on `Option` is called [`ok_or`](https://doc.rust-lang.org/std/option/enum.Option.html#method.ok_or), named after variant you want to get. In other words, "If it is some, give me an `ok` or this error".

More consistent names would be:

- `Option#some_or` and `Option#some_or_else`
- `Result#some` and `Result#none`

From these two choices, I much prefer the original naming on `Result` and hence would rather fix `Option` to say: `some_or` and `some_or_else`.

Going from here, the naming-scheme for `bool` -> `Option` could also follow this idea and be named after the variant that you already have. Unfortunately, `true` and `false` are keywords so we can't have function names `.true` and `.false`. What we could do is something like `map_true` and `map_false` with both of them taking a closure that produces the value for the `Some` variant:

```
true.map_true(|| 1) -> Some(1)
false.map_true(|| 1) -> None
false.map_true(|| 1) -> None
false.map_false(|| 1) -> Some(1)
```

**Comment by sanbox-irl** [comment]
`Option` and `Result` aren't getting their api changed anytime (ever actually I think? I think they're promised to never be anything worse than deprecated right?), so we have to live with those apis the way they are -- I think because of those inconsistencies, just looking at `Option` is probably the call

**Comment by thomaseizinger** [comment]
> `Option` and `Result` aren't getting their api changed anytime (ever actually I think? I think they're promised to never be anything worse than deprecated right?), so we have to live with those apis the way they are -- I think because of those inconsistencies, just looking at `Option` is probably the call

Oh yes, I wasn't trying to suggest that we need to change `Result` and `Option` :)

What I was trying to highlight is that even if they are not consistent, all the existing conversion functions mention either the "source" or "target" variant of the enum in their name and hence a name like "then" would not fit into this picture as it is lacking both.

**Comment by sanbox-irl** [comment]
Oh I see! My bad.

I agree with that conclusion -- I personally think the destination is just better, so I still prefer the simple `some` and `some_with`, but I'd be fine with your suggestions

**Comment by Alfriadox** [comment]
I'd be fine with the suggested `map_true`/`map_false`

**Comment by thomaseizinger** [comment]
> Oh I see! My bad.
> 
> I agree with that conclusion -- I personally think the destination is just better, so I still prefer the simple `some` and `some_with`, but I'd be fine with your suggestions

It is definitely more concise and avoids the "true"-is-a-keyword dilemma.
We should be aware that using the destination limits us to one conversion as we have to choose, what the function `some` does i.e. if it converts true -> Some, we can't also have one that goes from false -> Some and people will have to remember which one it is.

The "true" case seems more intuitive as also outlined further above on the topic of truthy and falsy boolean names.

This ambiguity is why I brought it up :)
I am fine either way though!

**Comment by Plecra** [comment]
If we're looking at consistent naming with the way `Option`s are "promoted" to `Result`s (Like we're promoting `bool`s to `Option`s), then I'd expect it to mirror the `Option::ok_or` API:

**`fn Option::ok_or<E>(self, err: E) -> Result<_, E>`**

which I read as "if we can (because the data is available), create an `Ok`, else create an `Err` with this".

With a `bool`, we only have the ability to create a `None`, and the caller needs to provide the data for the `Some` variant, making `Some` behave more like the `err` in the `Option` API, implying an API that I haven't seen brought up:

**`fn bool::none_or<T>(self, some: T) -> Option<T>`**

I'm not sure I actually *like* this version - it confuses the meaning of "or" and I expect it'd be confusing in code. That said, I wanted to mention it in case there's another name that uses `none` in the same way `ok_or` uses `ok`.

**Comment by mlvzk** [comment]
I'm confused why this is not stable yet, looks like a pretty small and simple change.

**Comment by sanbox-irl** [comment]
Because we can't agree on the name -- as withoutboats said above, this will probably become a standard part of a Rust programmer's toolkit, so we probably want to get it right. That being said...this has been going on for a really long time 😓

**Comment by KodrAus** [comment]
We discussed this at the recent Libs meeting and the only concrete argument we had against `then` and `then_some` was the length of the value-input vs closure-input being inconsistent with methods on `Result` and `Option`. The only alternative floated was:

```rust
impl bool {
    pub fn then_some<T>(self, t: T) -> Option<T>;
    pub fn and_then_some<T, F: FnOnce() -> T>(self, f: F) -> Option<T>;
}
```

It does seem like we've exhausted the design space for naming so I think now we just need to accept one and move on. Being a standard part of Rust's toolkit on the one hand means we do want to make a very deliberate choice in naming, but on the other means users are likely to learn them by-rote, possibly making them less reliant on any consistency relationships with other control-flow-esc-combinators.

@withoutboats did you have any preference on the names at this stage? Or should we FCP with just the `then` method?

**Comment by Boscop** [comment]
Btw, I think `and_then_some` as a name for `then` would be too long. 
Also it's confusing, it makes it sound like the closure should return an `Option<T>` because it sounds like it's a contraction of "bool to option" followed by `.and_then(f)`.

Also, in the meantime I've gotten used to the `then` and `then_some` methods.

But I think it would really make sense to have a way to express `and_then` too, instead of `b.then_some(()).and_then(|_| f)`. Maybe just `bool::and_then`? (In addition to `then` and `then_some`.)

**Comment by Alfriadox** [comment]
I still think `and_then` and `and_then_some` both have a lot of flaws and are arguably some of the worse names suggested in this thread. If we have the option to settle on anything else I think we should take it.

**Comment by ErichDonGubler** [comment]
https://github.com/rust-lang/rust/issues/64260#issuecomment-719501761 is interesting to me, because I would have thought `and_some` and `and_then_some` to be more consistent with the `Result` and `Option` APIs.

**Comment by m-ou-se** [comment]
I'm personally not a big fan of names like `map` for `bool`, simply because I don't see `bool` as a container or data structure like `Option`, `Result`, and `Vec` are. Although a boolean is in many ways similar to `Option<()>` or `Result<(), ()>`, the link with control flow ('if' / 'then') feels much more natural to me.

**Comment by thomaseizinger** [comment]
> I'm personally not a big fan of names like `map` for `bool`, simply because I don't see `bool` as a container or data structure like `Option`, `Result`, and `Vec` are. Although a boolean is in many ways similar to `Option<()>` or `Result<(), ()>`, the link with control flow ('if' / 'then') feels much more natural to me.

The reason I suggested `map` was because it does feel closer to the remaining functional-like APIs that can be chained. Those usually act like a pure pipeline of data transformations.

`if` on the other hand reminds me more of the `if_present` API in Java's `Optional`. This API takes a closure that _only_ runs if the `Optional` is present. I always disliked this API, it is pretty useless most of the time because you can't return anything from it.

However, I think the naming makes sense and is intuitive: Only run this closure if the present. It would be the same thing here: Only run this closure if the `bool` is `true`. In a way that is what is happening because we return a `Some` in that case but it is the data-transformation aspect that is intuitive to me when I read `if`. This might be due to the fact that I am not using `if` as an expression much in Rust :)

**Comment by Alfriadox** [comment]
I think it's a good idea to focus on the 'control flow' use of function. While `map*` is definitely closer to the functional, chain-able API on things like `Iterator`s, `Option`s, and `Result`s those all contain another type `T` and the associated methods are based around `map()`ing from `T` to `O`. Since `bool`s don't contain anything, `map*` naming is arguably weaker. While I don't personally dislike it all that much I can see how it's weaker than the naming based on the control flow. 

I'd also like to note that after thinking about it further, I think that `as_option` and `as_option_with` are just more verbose versions of `some` and `some_with`. If we choose to focus on on naming based on control-flow, these all also would be "out". 

Focusing on control flow narrows us down to two name pairs:
```rust 
fn if_true<T>(self, val: T) -> Option<T>;
fn on_true<T, F: FnOnce() -> T>(self, f: F) -> Option<T>;
```
or 
```rust
fn then_some<T>(self, val: T) -> T;
fn then<T, F: FnOnce() -> T>(self, f: F) -> Option<T>;
```

I think both of these are good candidates and while I personally am in favour of the first one, I would be able to settle for the second one if it's more popular.

**Comment by phimuemue** [comment]
Just my 2ct: The `bool`-to-`Option` approach covers manyof my use cases, but often need a bit more than a simple `bool` to `option`:

```
enum MyEnum {A(usize), B, C}
//...
if let A(n) = my_enum_value {
 Some(n)
} else {
 None
}
```

I.e. I need the if-then-else logic even if the condition is not a plain bool. Thus, I usually employ the following macro:

```
macro_rules! if_then_some {
    ($cond: expr, $val: expr) => {
        if $cond { Some($val) } else { None }
    };
    (let $pattern:pat = $expr: expr, $val: expr) => {
        if let $pattern = $expr { Some($val) } else { None }
    };
}
```

This covers simple bools and `if let`-expressions, as well. I do not know if a macro-based approach has been discussed and discarded - I just wanted to make sure that it does not go unnoticed in case someone else finds it useful.

**Comment by slerpyyy** [comment]
One thing I would like to add:

The words `then` and `else` should only be used to indicate lazy evaluation.

This is a convention that goes back to SML and OCaml, which use the `if ... then ... else ...` syntax for conditionals. SML also uses the keywords `and`/`or` for eagerly evaluated logic and `andalso`/`orelse` for lazily evaluated logic. (I have no idea why it's not `andthen`). This pattern is already reflected in many existing functions in Rust; see for example `{Option, Result}::{and, and_then, or, or_else, map_or, map_or_else, unwrap_or, unwrap_or_else}` as well as `Option::{ok_or, ok_or_else}`.

**Comment by Boscop** [comment]
After thinking about this more, I now think that the names `some` and `some_with` make the most sense.
(I was kinda getting used to `then` and `then_some`, but I don't like that the one that takes the value has the longer name, it's inconsistent.)

And I really think we should also have a method that take a closure that returns an `Option` in the success case, like [`and_option_from`](https://docs.rs/boolinator/2.4.0/boolinator/trait.Boolinator.html#tymethod.and_option_from). Maybe it could be called `option_with`? The name would be consistent with `some` and `some_with`!

(Note that we don't need a method for [`and_option`](https://docs.rs/boolinator/2.4.0/boolinator/trait.Boolinator.html#tymethod.and_option) because that's covered by `opt.filter(cond)`.)

---

@phimuemue I proposed adding such a macro (`match_map`) here: https://github.com/rust-lang/rfcs/issues/2960
Your example would be written as: `match_map!(my_enum_value, A(n) => n)`.
The macro also supports match-guards.

**Comment by Alfriadox** [comment]
> After thinking about this more, I now think that the names `some` and `some_with` make the most sense.
> (I was kinda getting used to `then` and `then_some`, but I don't like that the one that takes the value has the longer name, it's inconsistent.)
> 
> And I really think we should also have a method that take a closure that returns an `Option` in the success case, like [`and_option_from`](https://docs.rs/boolinator/2.4.0/boolinator/trait.Boolinator.html#tymethod.and_option_from). Maybe it could be called `option_with`? The name would be consistent with `some` and `some_with`!
> 
I could see `some` and `some_with`. 
I don't think we need the other one there though as it should be covered by `Option::and_then`. 
For example:
```rust
let opt_fn = |a: u32| (a < 10).some(a).and_then(|n| (n >= 1).some(n*2));
assert_eq!(opt_fn(0), None);
assert_eq!(opt_fn(10), None);
assert_eq!(opt_fn(5), Some(10));
```

**Comment by KodrAus** [comment]
We've had lot of participation here with a lot of perspectives on how to name these functions being discussed! It doesn't seem like this latest round has really brought anything new to the table as far as naming goes and we haven't produced anything universally preferred to the current API of:

```rust
impl bool {
    pub fn then_some<T>(self, t: T) -> Option<T>;

    pub fn then<T, F>(self, f: F) -> Option<T> where F: FnOnce() -> T;
}
```

I don't think more discussion is likely to change this, and an alternative to `then` that returns an `Option<T>` could be named to fit within this scheme, so in the interests of unblocking these APIs for stable users and providing some closure on this whole journey we've been on, I'm going to propose we stabilize the status-quo of `then` and `then_some`.

@rfcbot fcp merge

**Comment by KodrAus** [comment]
Looks like `rfcbot` can't handle two FCPs in the same issue at the same time 😄 

@rfcbot cancel

**Comment by KodrAus** [comment]
@rfcbot fcp merge

**Comment by withoutboats** [comment]
@rfcbot concern stabilize-only-then

> @withoutboats did you have any preference on the names at this stage? Or should we FCP with just the then method?

Sorry for not responding sooner. For the reasons previously stated, I would prefer to stabilize only `then`. I don't think any other names than `then` and `then_some` should be considered at this point, but I'd like to have `then` on stable so we have more experience before considering if `then_some` as sugar over `then` carries its weight.

**Comment by varkor** [comment]
> I'd like to have then on stable so we have more experience before considering if then_some as sugar over then carries its weight.

I'll just note that `then_some` is already used frequently throughout rustc. In fact, `then_some` occurs more times (43 times) than `then` (16 times). Personally, I think this already demonstrates its usefulness.

**Comment by BurntSushi** [comment]
I personally find these combinators to be less clear than an if statement, and they don't really save much typing. So I doubt I'd find myself using them and would probably oppose them in PR reviews for my crates. With that said, I don't feel strongly enough to mount an opposition to them. I don't think my clarity argument is that strong IMO, and is probably closer to a personal preference.

I also think that if we stabilize one of them, we should probably just stabilize both. The pattern of having one combinator accept a value and one accept a closure is pretty common on our `Result` and `Option` types, so it seems like it fits well with the pattern of our other APIs. On the other hand, I'm rarely opposed to taking the conservative path either.

**Comment by Coder-256** [comment]
I hate to be another random person contributing to the bikeshedding, during a FCP vote; I especially hate to hijack the thread with more unrelated, RFC-length bikeshedding; and I know that admitting my admittedly grievous sin doesn't make it any better, but I wanted to mention something that I think hasn't been said before.

IMO the general pattern of double combinators where one takes a value and another takes a closure, is a very bad design. This will get a bit off-topic but I'd like to explain broadly why I think this is a bad choice, not just for `then`/`then_some`.

Here are examples of the type of combinator that I am talking about:

| Value                   | Closure                      |
| ----------------------- | ---------------------------- |
| `Option::and`           | `Option::and_then`           |
| `Option::get_or_insert` | `Option::get_or_insert_with` |
| `Option::map_or`        | `Option::map_or_else`        |
| `Option::ok_or`         | `Option::ok_or_else`         |
| `Option::or`            | `Option::or_else`            |
| `Option::unwrap_or`     | `Option::unwrap_or_else`     |
| `Result::and`           | `Result::and_then`           |
| `Result::map_or`        | `Result::map_or_else`\*      |
| `Result::or`            | `Result::or_else`            |
| `Result::unwrap_or`     | `Result::unwrap_or_else`\*   |
| _`bool::then_some`_     | _`bool::then`_               |

<sup>\* = The closure form can also access the error (yet another benefit)</sup>

<sup>Side thought: why is there no `Result::get_or_insert{,_with}`?</sup>

Let's use `Option::unwrap_or` and `Option::unwrap_or_else` for example.

It seems to be considered idiomatic to use `foo.unwrap_or(bar)` for cheap values and and `foo.unwrap_or_else(|| expensive_call())` for expensive calls. However, it's not clear why the former is any better than the latter. In other words, it's not clear why the former version _exists_. As far as I am aware, the only benefits for using the former instead of `foo.unwrap_or_else(|| bar)` are:

1. It's shorter. I imagine this is the main reason why most people tend towards it when possible.
2. It seems more "fitting" than the closure version; you pick it when you don't need lazy evaluation.
3. You want eager evaluation.

However, the first two points are actually dependent on the third, because they are only valid if we have two sets of combinators. Leaving us the third.

Stop. Please think long and hard. Have you, and when have you, ever actually needed eager evaluation with any of these functions? Now... is it worth doubling them all?

Is there really any reason for the entire left column in the table above?

IMO This pattern only serves to create confusion, code duplication, and an _unnecessary_ amount of ways to do the same thing. I'm not sure how or why this pattern started (maybe performance before closures were well-optimized?) but it seems unnecessary today.

What I consider to be an infinitely better design is: rename everything in the second column to the name in first column (e.g. `unwrap_or_else` → `unwrap_or`), remove the latter (in an edition), and add a macro for when you _really_ need eager evaluation. Something like this:

```rust
macro_rules! eager {
    ($e:expr) => {
        {
            let x = $e;
            move || x
        }
    }
}

// Old:
option.unwrap_or_else(|| slow() + 123)
option.unwrap_or(slow() + 123)
option.unwrap_or(value)

// New:
option.unwrap_or(|| slow() + 123)
option.unwrap_or(eager!(slow() + 123))
option.unwrap_or(|| value)
```

If we replaced the non-closure methods, then:

- For new users, there is only one method to choose from.
- Anything you want to do the old way, you can still do the new way. If you really need eager evaluation (which I strongly feel is the exception, not the norm) then you can use `eager!()`.
- The methods more closely mirror the normal `if`/`else` control flow. They more directly translate from the `match` block they elide.
- You only ever need to add `||` to pass a closure instead of an eagerly-evaluated expression (99% sure that no cases would need brackets).
- Performance when converting to a closure (assuming that these combinators are all inline):
  - For expressions with no side effects that only call inline functions (like `u8::add`), the closure should be zero-cost when optimized, so there should be no difference (although this must be tested).
  - For expressions with side effects, or closures that call non-inline functions... I'd imagine that users rarely actually want eager evaluation! In most cases, performance should be unchanged if not increase, otherwise you can use `eager!()` to get back the old behavior.
- IMO this design is much clearer and only slightly more verbose than the current combinators (and less verbose if you're already using a closure).
- For Ferris' sake it will help the `bool` bikeshedding!

For prior art, see [Swift autoclosures](https://docs.swift.org/swift-book/LanguageGuide/Closures.html#ID543).

---

If you're still not convinced, well I decided to analyze all uses of `then_some` in the latest master rustc (dc06a36074f04c6a77b5834f2950011d49607898). There were 39, excluding the definition/doctests. I checked, and the argument to every single one was either:

- An existing variable.
- A type constructor with existing variables.
- A method chain that delegates to one of the above (including `vec![]`).

(with [one exception](https://github.com/rust-lang/rust/blob/dc06a36074f04c6a77b5834f2950011d49607898/compiler/rustc_codegen_llvm/src/common.rs#L226)).

None of them benefit from eager evaluation.

The reason I am bringing this up now is that in lieu of the major changes that I've proposed, I at least wouldn't like to stabilize what I consider to be another mistake.

TL;DR `b.then_some(x)` is bad. Just use `b.then(|| x)`.

**Comment by withoutboats** [comment]
> I also think that if we stabilize one of them, we should probably just stabilize both. The pattern of having one combinator accept a value and one accept a closure is pretty common on our Result and Option types, so it seems like it fits well with the pattern of our other APIs.

My concern is that this *doesn't* fit the pattern of our other APIs:

* The extended name is the one that takes a value, rather than the one that takes a closure.
* The by-value combinator includes `then` in its name, weakening the association between the word "then" and taking a closure that is otherwise common across our combinators.

As an additional concern with this pair of names, I'll point out:

* The longer name's extension, `some`, does not provide much clarity about which is which, because both apply `Some` to their argument. It might even confuse users about the difference in their signature (thinking that `then` takes `FnOnce() -> Option<T>` fo rexample).

However, I think the search for a better pair of names without these concerns has proved fruitless.

As was pointed out in the comment that has been marked off-topic, the pair is technically redundant: all you need is the lazy version. And all of these concerns are obliviated by removing the non-lazy `then_some` combinator. So my solution is to merge `then` and then see if the concerns about not having `then_some` actually outweigh the concerns with having it.

@varkor's statistics are not a compelling argument to me: since `then_some` is strictly less powerful than `then`, all of those uses of `then_some` can be trivially replaced by `then`. The question is not "are there many use cases where you don't need lazy evaluation" (seems obvious to me the answer will be "yes"); the question is "would the benefits of having an eager combinator here outweigh the costs of introducing some naming inconsistency?"

**Comment by m-ou-se** [comment]
> FWIW from my experiments on nightly I think then is a very clear name, I think the API is very useful, and I can already tell I will use it all the time once its stable!

In that case, how about an FCP for only `then`? Seems like that's the only way right now to get this moving forward. Then we can leave `then_some` unstable for now and see if anybody feels the need to get that stabilized later (or remove it otherwise).

**Comment by withoutboats** [comment]
> In that case, how about an FCP for only then? Seems like that's the only way right now to get this moving forward. Then we can leave then_some unstable for now and see if anybody feels the need to get that stabilized later (or remove it otherwise).

I'm fully on board with that! That's what I hoped to convey by checking my box and filing the concern at the same time.

**Comment by Boscop** [comment]
@withoutboats Why not use the names `some` and `some_with`?

I think it makes sense to have an eager version because the closure won't be optimized away so it's not a zero cost abstraction.

**Comment by m-ou-se** [comment]
> because the closure won't be optimized away

@Boscop Why wouldn't it be optimized away? I can't find any example (where the expression doesn't have side effects / is constant) where `.then(|| ..)` and `.then_some(..)` result in different output on any opt level except 0. ([Example](https://godbolt.org/z/qY7P7G).)

**Comment by Boscop** [comment]
@m-ou-se Ah ok, I thought it won't optimize it. But yeah, if it always does, then it makes sense to only have the closure version. (Although it still would be nice to have a way to get the optimized version in debug builds.)

@Alfriadox 
> I don't think we need the other one there though as it should be covered by `Option::and_then`.

I mean this pattern: `if b { f() } else { None }` can't be expressed with `and_then` because there is no initial `Option`! 
And `b.then(f)` can't be used because `f()` returns a `Option`!
So in addition to `then` we should have a method like this, named `and_then` or `and_option`.

**Comment by Plecra** [comment]
The "original" option can be created using `then`:

```rust
b.then(()).and_then(f)
```

**Comment by m-ou-se** [comment]
Or `b.then(f).flatten()`.

**Comment by m-ou-se** [comment]
@KodrAus @sfackler do your checkmarks still count if this is only for `then`? (I'm assuming @BurntSushi's checkmark does, looking at the thumbs-up-reactions to the comments suggesting this.) If so, we can resolve @withoutboats' concern and continue with the FCP.

**Comment by KodrAus** [comment]
@m-ou-se Yep! I'm also ok with stabilizing just `then`.

**Comment by sfackler** [comment]
I am on board with that as well.

**Comment by withoutboats** [comment]
@fcbot resolve stabilize-only-then

We agreed in the libs team today to stabilize just `then` for now.

**Comment by withoutboats** [comment]
@rfcbot resolve stabilize-only-then

**Comment by varkor** [comment]
I've opened a stabilisation PR for `then`: https://github.com/rust-lang/rust/pull/79299.

**Comment by jplatte** [comment]
Stabilized; I think this tracking issue can be closed 🙂

**Comment by CryZe** [comment]
Only `then` got stabilized, `then_some` is still being tracked.

**Comment by m-ou-se** [comment]
It might be a good idea to move `then_some` to a new tracking issue and close this one.

**Comment by varkor** [comment]
I've updated the title and description. I don't think there's any need to create a new issue: that just causes unnecessary churn.

**Comment by slerpyyy** [comment]
Does that mean we're still considering stabilizing `then_some` after all?
I thought we decided that adding `then_some` was redundant, since `then` can already do everything `then_some` can and more.

**Comment by varkor** [comment]
@slerpyyy: yes, see https://github.com/rust-lang/rust/issues/64260#issuecomment-723161462. There was a desire to see how useful `then_some` was in practice as a shorthand. This pattern (taking a value vs taking a closure) is already common throughout the standard library.

**Comment by KodrAus** [comment]
Closing this one in favour of #80967, so that we'll get a chance to issue a fresh FCP if we decide to in the future.

**Comment by matklad** [comment]
As an experience report, I've recently found roughly the following code in the wild: 

```rust
(!is_uppercase)
    .then(|| ())
    .ok_or_else(|| ParseError(ParseErrorKind::Invalid, value.to_string()))
```

It takes some mental puzzle-solving to figure out whether `is_uppercase` results in error or not.

Don't think this is actionable in any way, but it's a useful real-world specimen to ponder.

**Comment by thomaseizinger** [comment]
I think that is a good example that not everything is better with combinators :)

I haven't decipered 100% it but I would write this as:

```rust
if !is_uppercase {
    return Err(ParseError(...));
}
```

Personally, I've been using `.then()` primarily with `.filter_map`. I think it is a fantasic fit for that.

**Comment by Lucretiel** [comment]
Agree. I've seen some similar code in the wild. I think that *usually* `.then()` is only appropriate in cases where an `Option` (or `Result`, via `ok_or_else`) is actually in play; `.then()` followed immediately by `*_or_else` is probably an antipattern (because you're needlessly round-tripping through an `Option`. I'd love to see a clippy lint to this effect.

**Comment by bcpeinhardt** [comment]
Hey there! I've just come from fixing a bug I created. What's the reasoning behind `false.then_some(expr())` not short circuiting/ would there be support for a short circuiting verison, like `then_eval_some`? I realize you can use `then` and it will effectively short circuit because the closure won't be called, but it's a little clunky with methods that return a Result because you end up transposing in order to use `?`. For reference

The original
```
let mut initializer = None;
if self.advance_on(TokenType::Equal) {
    initializer = Some(self.expression()?);
}
```

The bug I created because I didn't realize false.then wouldn't short circuit.
```
let initializer = self
    .advance_on(TokenType::Equal)
    .then_some(self.expression()?);
```

The current working version which does short circuit
```
let initializer = self
    .advance_on(TokenType::Equal)
    .then(|| self.expression()).transpose()?;
```

I feel like if you wanted to evaluate the provided value no matter what, the clearest code would be
```
let expr = self.expression()?;
let initializer = self.advance_on(TokenType::Equal).then_some(expr);
```

As a side note, I also find the naming strange. When I see two methods called then and then_some, I think one of them wraps the provided value in an option for you and one doesn't, not that one takes a closure and one takes a value.
